crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+            opts,
+            inflight: Default::default(),
+            rx,
+            tx,
+        }
+    }
+
+    /// Returns a handle that can be used by callers to enqueue new records.
+    pub fn handle(&self) -> SessionHandle {
+        SessionHandle {
+            tx: self.tx.clone(),
+        }
+    }
+}
+
+impl<T> PartitionSession<T>
+where
+    T: TransportConnect,
+{
+    /// Runs the session state machine until shut down, reacting to cancellation and connection errors.
+    pub async fn start(mut self, ctx: CancellationToken) {
+        let mut state = SessionState::Connecting {
+            retry: self.opts.connect_retry_policy.clone().into_iter(),
+        };
+
+        debug!(
+            partition_id = %self.partition,
+            "Starting ingress partition session",
+        );
+
+        loop {
+            state = match state {
+                SessionState::Connecting { retry } => self.connect(&ctx, retry).await,
+                SessionState::Connected { connection } => self.connected(&ctx, connection).await,
+                SessionState::Disconnected => SessionState::Connecting {
+                    retry: self.opts.connect_retry_policy.clone().into_iter(),
+                },
+                SessionState::Shutdown => {
+                    self.rx.close();
+                    break;
+                }
+            }
+        }
+    }
+
+    async fn connect(
+        &self,
+        ctx: &CancellationToken,
+        mut retry: RetryIter<'static>,
+    ) -> SessionState {
+        let Some(node_id) = self.partition_routing.get_node_by_partition(self.partition) else {
+            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+
+            return SessionState::Connecting { retry };
+        };
+
+        tokio::select! {
+            result = self
+                .networking
+                .get_connection(node_id, Swimlane::IngressData) => {
+                    match result {
+                        Ok(connection) => SessionState::Connected { connection },
+                        Err(ConnectError::Shutdown(_)) => SessionState::Shutdown,
+                        Err(err) => {
+                            debug!("Failed to connect to node {node_id}: {err}");
+                            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+                            SessionState::Connecting { retry }
+                        }
+                    }
+            }
+            _ = ctx.cancelled() => {
+                SessionState::Shutdown
+            }
+        }
+    }
+
+    /// Re-sends all inflight batches after a connection is restored.
+    async fn replay(&mut self, connection: &Connection) -> Result<(), ConnectionClosed> {
+        //todo(azmy): to avoid all the inflight batches again and waste traffic
+        //maybe test the connection first by sending an empty batch and wait for response
+        //before proceeding?
```

Comment:
```suggestion
        // todo(azmy): to avoid all the inflight batches again and waste traffic
        //  maybe test the connection first by sending an empty batch and wait for response
        //  before proceeding?
```